src/http/server/handlers/static_file_handler/cache.cr

```diff
@@ -0,0 +1,48 @@
+class HTTP::StaticFileHandler::Cache
+  HTTP_IF_MODIFIED_SINCE = "If-Modified-Since"
+  HTTP_IF_NONE_MATCH     = "If-None-Match"
+  HTTP_ETAG              = "Etag"
+  HTTP_LAST_MODIFIED     = "Last-Modified"
+
+  def self.add_headers(context : HTTP::Server::Context, file_path : String) : Nil
+    context.response.headers[HTTP_ETAG] = etag(file_path)
+
+    last_modified = modification_time(file_path)
+    context.response.headers[HTTP_LAST_MODIFIED] = HTTP.rfc1123_date(last_modified)
+  end
+
+  def self.match?(context : HTTP::Server::Context, file_path : String) : Bool
+    etag?(context, file_path) || last_modified?(context, file_path)
+  end
+
+  private def self.etag(file_path)
+    %{W/"#{File.info(file_path).modification_time.epoch.to_s}"}
+  end
+
+  private def self.modification_time(file_path)
+    File.info(file_path).modification_time
+  end
+
+  private def self.etag?(context : HTTP::Server::Context, file_path : String) : Bool
+    if_none_match = context.request.headers[HTTP_IF_NONE_MATCH]?
+    return false unless if_none_match
+
+    if_none_match == etag(file_path)
```

Comment:
You don't need to recompute the etag, it's already available as `context.response.headers[HTTP_ETAG]`. Ditto below.